File: app/src/main/java/com/example/furiganakeyboard/ime/RomajiCompositionEditor.kt
package com.example.furiganakeyboard.ime

/**
 * Pure romaji editing policy around an authoritative [RomajiCompositionCursor].
 *
 * The service owns Android rendering and candidate refresh. This class owns only the transformation
 * from incremental raw keys to a pending middle range or resolved kana, so right-side composition
 * text can never be rebuilt from an end-only buffer.
 */
internal class RomajiCompositionEditor(
    private val cursor: RomajiCompositionCursor,
) {
    fun append(rawUnit: String): CompositionCursorMutation {
        require(rawUnit.isNotEmpty()) { "rawUnit must not be empty" }
        return replaceRaw(cursor.pendingEdit?.raw.orEmpty() + rawUnit)
    }

    fun deleteBeforeCursor(): CompositionCursorMutation {
        val pending = cursor.pendingEdit
        return if (pending != null) {
            replaceRaw(RomajiKanaConverter.deleteLastUnit(pending.raw))
        } else {
            cursor.deleteGraphemeBeforeCursor()
        }
    }

    private fun replaceRaw(raw: String): CompositionCursorMutation {
        val pendingBefore = cursor.pendingEdit
        if (raw.isEmpty()) {
            return if (pendingBefore == null) {
                cursor.deleteGraphemeBeforeCursor()
            } else {
                cursor.resolvePendingRomaji("")
            }
        }

        val converted = RomajiKanaConverter.convert(raw)
        return when {
            converted.hasUnresolvedInput ->
                cursor.replacePendingRomaji(converted.displayText, raw)
            pendingBefore != null ->
                cursor.resolvePendingRomaji(converted.kana)
            else ->
                cursor.insertResolvedAtCursor(converted.kana)
        }
    }
}

File: ios/Sources/KeyboardExtension/RomajiCompositionEditor.swift
import Foundation
#if REPOSITORY_TEST_HOST
@testable import FuriganaKeyboard
#endif

/// Pure romaji editing policy around an authoritative `RomajiCompositionCursor`.
///
/// The keyboard controller owns rendering and candidate refresh. This type owns only the
/// transformation from incremental raw keys to a pending middle range or resolved kana, so
/// text right of the cursor can never be rebuilt from an end-only buffer.
struct RomajiCompositionEditor {
    private let cursor: RomajiCompositionCursor

    init(cursor: RomajiCompositionCursor) {
        self.cursor = cursor
    }

    func append(_ rawUnit: String) -> CompositionCursorMutation {
        precondition(!rawUnit.isEmpty, "rawUnit must not be empty")
        return replaceRaw((cursor.pendingEdit?.raw ?? "") + rawUnit)
    }

    func deleteBeforeCursor() -> CompositionCursorMutation {
        guard let pending = cursor.pendingEdit else { return cursor.deleteGraphemeBeforeCursor() }
        return replaceRaw(RomajiKanaConverter.deleteLastUnit(pending.raw))
    }

    private func replaceRaw(_ raw: String) -> CompositionCursorMutation {
        let hadPending = cursor.pendingEdit != nil
        guard !raw.isEmpty else {
            return hadPending
                ? cursor.resolvePendingRomaji("")
                : cursor.deleteGraphemeBeforeCursor()
        }
        let converted = RomajiKanaConverter.convert(raw)
        if converted.hasUnresolvedInput {
            return cursor.replacePendingRomaji(display: converted.displayText, raw: raw)
        }
        return hadPending
            ? cursor.resolvePendingRomaji(converted.kana)
            : cursor.insertResolvedAtCursor(converted.kana)
    }
}
